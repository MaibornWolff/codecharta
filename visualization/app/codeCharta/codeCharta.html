<div id="hud" class="container">

    <div class="row">

        <div class="col l3 m6 s12 col-backdrop">
            <legend-panel-component></legend-panel-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <revision-chooser-component></revision-chooser-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <file-chooser-panel-directive></file-chooser-panel-directive>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <settings-panel-component></settings-panel-component>
        </div>

    </div>

    <div class="row bottom-row">

        <div class="col l3 m6 s12 col-backdrop">
            <detail-panel-component></detail-panel-component>
        </div>

    </div>

    <div id="titlebox">
        <a class="mwlink" href="https://www.maibornwolff.de" target="_blank"><img src="./logo.png" alt="MaibornWolff"></a>
        <h1 id="title"><span>CodeCharta {{$ctrl.pkg.version}}</span></h1>
    </div>

</div>

<code-map-component></code-map-component>