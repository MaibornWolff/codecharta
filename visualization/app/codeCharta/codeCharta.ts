"use strict";

import "./core/core.module";
import "./ui/ui";

import {codeChartaComponent, CodeChartaController} from "./codeCharta.component";

import angular from "angular";

angular.module(
    "app.codeCharta",
    ["app.codeCharta.core", "app.codeCharta.ui"]
);

angular.module("app.codeCharta").component(
    codeChartaComponent.selector,
    codeChartaComponent
);


