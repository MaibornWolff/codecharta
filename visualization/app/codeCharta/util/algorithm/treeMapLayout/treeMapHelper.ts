import { getMapResolutionScaleFactor, getMarkingColor, isLeaf } from "../../codeMapHelper"
import { CcState, CodeMapNode, ColorMode, Node } from "../../../codeCharta.model"
import { Vector3 } from "three"
import { CodeMapBuilding } from "../../../ui/codeMap/rendering/codeMapBuilding"
import { HierarchyRectangularNode } from "d3-hierarchy"
import { searchedNodePathsSelector } from "../../../state/selectors/searchedNodes/searchedNodePaths.selector"
import { gradientCalculator } from "../../color/gradientCalculator"
import {
    MetricMinMax,
    selectedColorMetricDataSelector
} from "../../../state/selectors/accumulatedData/metricData/selectedColorMetricData.selector"

export const treeMapSize = 250

const FOLDER_HEIGHT = 2
const MIN_BUILDING_HEIGHT = 2
const HEIGHT_VALUE_WHEN_METRIC_NOT_FOUND = 0

function countNodes(node: { children?: CodeMapNode[] }) {
    let count = 1
    if (node.children) {
        for (const child of node.children) {
            count += countNodes(child)
        }
    }
    return count
}

// TODO this function exists twice in the code - please refactor it.
function calculateSize(node: CodeMapNode, metricName: string) {
    let totalSize = node.attributes[metricName] || 0

    if (totalSize === 0 && node.children && node.children.length > 0) {
        for (const child of node.children) {
            totalSize += calculateSize(child, metricName)
        }
    }
    return totalSize
}

function buildingArrayToMap(highlighted: CodeMapBuilding[]) {
    const geomMap: Map<number, CodeMapBuilding> = new Map()

    for (const building of highlighted) {
        geomMap.set(building.id, building)
    }

    return geomMap
}

function buildRootFolderForFixedFolders(map: CodeMapNode, heightScale: number, state: CcState, isDeltaState: boolean) {
    const flattened = isNodeFlat(map, state)
    const height = FOLDER_HEIGHT
    const width = 100
    const length = 100

    return {
        name: map.name,
        id: 0,
        width,
        height,
        length,
        depth: 0,
        x0: 0,
        z0: 0,
        y0: 0,
        isLeaf: false,
        attributes: map.attributes,
        edgeAttributes: map.edgeAttributes,
        deltas: map.deltas,
        heightDelta: (map.deltas?.[state.dynamicSettings.heightMetric] ?? 0) * heightScale,
        visible: isVisible(map, false, state, flattened),
        path: map.path,
        link: map.link,
        markingColor: getMarkingColor(map, state.fileSettings.markedPackages),
        flat: false,
        color: getBuildingColor(map, state, selectedColorMetricDataSelector(state), isDeltaState, flattened),
        incomingEdgePoint: getIncomingEdgePoint(width, height, length, new Vector3(0, 0, 0), treeMapSize),
        outgoingEdgePoint: getOutgoingEdgePoint(width, height, length, new Vector3(0, 0, 0), treeMapSize)
    } as Node
}

function buildNodeFrom(
    squaredNode: HierarchyRectangularNode<CodeMapNode>,
    heightScale: number,
    maxHeight: number,
    state: CcState,
    isDeltaState: boolean
): Node {
    const mapSizeResolutionScaling = getMapResolutionScaleFactor(state.files)
    const { x0, x1, y0, y1, data } = squaredNode
    const isNodeLeaf = isLeaf(squaredNode)
    const flattened = isNodeFlat(data, state)
    const heightValue = getHeightValue(state, data, maxHeight, flattened)
    const depth = data.path.split("/").length - 2
    const height = isNodeLeaf ? resolveHeightValue(heightValue, heightScale, data, state) * mapSizeResolutionScaling : FOLDER_HEIGHT
    const width = x1 - x0
    const length = y1 - y0
    const z0 = squaredNode.depth * FOLDER_HEIGHT
    const heightDelta = (data.deltas?.[state.dynamicSettings.heightMetric] ?? 0) * heightScale * mapSizeResolutionScaling
    const edgePointHeight = height + (heightDelta < 0 ? Math.abs(heightDelta) : 0)

    return {
        name: data.name,
        id: data.id,
        width,
        height,
        length,
        depth,
        mapNodeDepth: squaredNode.depth,
        x0,
        z0,
        y0,
        isLeaf: isNodeLeaf,
        attributes: data.attributes,
        edgeAttributes: data.edgeAttributes,
        deltas: data.deltas,
        heightDelta,
        visible: isVisible(data, isNodeLeaf, state, flattened),
        path: data.path,
        link: data.link,
        markingColor: getMarkingColor(data, state.fileSettings.markedPackages),
        flat: flattened,
        color: getBuildingColor(data, state, selectedColorMetricDataSelector(state), isDeltaState, flattened),
        incomingEdgePoint: getIncomingEdgePoint(width, edgePointHeight, length, new Vector3(x0, z0, y0), treeMapSize),
        outgoingEdgePoint: getOutgoingEdgePoint(width, edgePointHeight, length, new Vector3(x0, z0, y0), treeMapSize)
    }
}

export function getHeightValue(state: CcState, squaredNode: CodeMapNode, maxHeight: number, flattened: boolean) {
    const mapSizeResolutionScaling = getMapResolutionScaleFactor(state.files)

    if (flattened) {
        return MIN_BUILDING_HEIGHT
    }

    let heightValue = squaredNode.attributes[state.dynamicSettings.heightMetric] || HEIGHT_VALUE_WHEN_METRIC_NOT_FOUND
    heightValue *= mapSizeResolutionScaling

    const heightMetric = state.dynamicSettings.heightMetric
    const attributeDescriptors = state.fileSettings.attributeDescriptors
    const isAttributeDirectionInversed = attributeDescriptors[heightMetric]?.direction === 1

    if (isAttributeDirectionInversed) {
        if (state.appSettings.invertHeight) {
            return heightValue
        }
        return maxHeight - heightValue
    }
    if (state.appSettings.invertHeight) {
        return maxHeight - heightValue
    }
    return heightValue
}

export function resolveHeightValue(heightValue: number, heightScale: number, data: CodeMapNode, state: CcState): number {
    const minimalHeight = data.deltas?.[state.dynamicSettings.heightMetric] ? 0 : MIN_BUILDING_HEIGHT
    return Math.max(Math.abs(heightScale * heightValue), minimalHeight)
}

export function isVisible(squaredNode: CodeMapNode, isNodeLeaf: boolean, state: CcState, flattened: boolean) {
    if (squaredNode.isExcluded || (isNodeLeaf && state.appSettings.hideFlatBuildings && flattened)) {
        return false
    }

    if (state.dynamicSettings.focusedNodePath.length > 0) {
        return squaredNode.path.startsWith(state.dynamicSettings.focusedNodePath[0])
    }

    return true
}

export function getIncomingEdgePoint(width: number, height: number, length: number, vector: Vector3, mapSize: number) {
    if (width > length) {
        return new Vector3(vector.x - mapSize + width / 4, vector.y + height, vector.z - mapSize + length / 2)
    }
    return new Vector3(vector.x - mapSize + width / 2, vector.y + height, vector.z - mapSize + length / 4)
}

export function getOutgoingEdgePoint(width: number, height: number, length: number, vector: Vector3, mapSize: number) {
    if (width > length) {
        return new Vector3(vector.x - mapSize + 0.75 * width, vector.y + height, vector.z - mapSize + length / 2)
    }
    return new Vector3(vector.x - mapSize + width / 2, vector.y + height, vector.z - mapSize + 0.75 * length)
}

export function isNodeFlat(codeMapNode: CodeMapNode, state: CcState) {
    if (codeMapNode.isFlattened) {
        return true
    }

    const searchedNodePaths = searchedNodePathsSelector(state)

    if (searchedNodePaths && state.dynamicSettings.searchPattern?.length > 0) {
        return searchedNodePaths.size === 0 || isNodeNonSearched(codeMapNode, state)
    }

    if (state.appSettings.showOnlyBuildingsWithEdges && state.fileSettings.edges.some(edge => edge.visible)) {
        return nodeHasNoVisibleEdges(codeMapNode, state)
    }

    return false
}

function nodeHasNoVisibleEdges(codeMapNode: CodeMapNode, state: CcState) {
    return (
        codeMapNode.edgeAttributes[state.dynamicSettings.edgeMetric] === undefined ||
        !state.fileSettings.edges.some(edge => codeMapNode.path === edge.fromNodeName || codeMapNode.path === edge.toNodeName)
    )
}

function isNodeNonSearched(squaredNode: CodeMapNode, state: CcState) {
    const searchedNodePaths = searchedNodePathsSelector(state)
    return !searchedNodePaths.has(squaredNode.path)
}

export function getBuildingColor(
    node: CodeMapNode,
    { appSettings, dynamicSettings }: CcState,
    nodeMetricDataRange: MetricMinMax,
    isDeltaState: boolean,
    flattened: boolean
) {
    const { mapColors } = appSettings

    if (isDeltaState) {
        return mapColors.base
    }
    const metricValue = node.attributes[dynamicSettings.colorMetric]

    if (metricValue === undefined) {
        return mapColors.base
    }
    if (flattened) {
        return mapColors.flat
    }

    const { colorRange, colorMode } = dynamicSettings

    if (dynamicSettings["colorMetric"] === "unary") {
        return mapColors.positive
    }

    if (colorMode === ColorMode.absolute) {
        if (metricValue < colorRange.from || colorRange.from === nodeMetricDataRange.maxValue) {
            return mapColors.positive
        }
        if (metricValue < colorRange.to || colorRange.to === nodeMetricDataRange.maxValue) {
            return mapColors.neutral
        }
        return mapColors.negative
    }

    if (colorMode === ColorMode.trueGradient) {
        return gradientCalculator.getColorByTrueGradient(mapColors, colorRange, nodeMetricDataRange, metricValue)
    }

    if (colorMode === ColorMode.focusedGradient) {
        return gradientCalculator.getColorByFocusedGradient(mapColors, colorRange, nodeMetricDataRange, metricValue)
    }

    return gradientCalculator.getColorByWeightedGradient(mapColors, colorRange, nodeMetricDataRange, metricValue)
}

export const TreeMapHelper = {
    countNodes,
    buildingArrayToMap,
    buildRootFolderForFixedFolders,
    calculateSize,
    buildNodeFrom,
    isNodeFlat,
    resolveHeightValue,
    getHeightValue,
    FOLDER_HEIGHT,
    MIN_BUILDING_HEIGHT,
    HEIGHT_VALUE_WHEN_METRIC_NOT_FOUND
}
