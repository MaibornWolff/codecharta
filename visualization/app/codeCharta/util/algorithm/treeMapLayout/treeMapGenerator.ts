import { hierarchy, HierarchyNode, HierarchyRectangularNode, treemap } from "d3-hierarchy"
import { TreeMapHelper, treeMapSize } from "./treeMapHelper"
import { CodeMapNode, DynamicSettings, Node, NodeMetricData, State } from "../../../codeCharta.model"
import { getMapResolutionScaleFactor, isLeaf } from "../../codeMapHelper"
import { calculatePaddingBasedOnBuildingArea } from "./paddingCalculator"
import { getChildrenAreaValues, getSmallestValueOrSmallestDifference } from "./modifiedTreeMapHelperFunctions"
import { calculateTotalNodeArea } from "./nodeAreaCalculator"

export type SquarifiedTreeMap = { treeMap: HierarchyRectangularNode<CodeMapNode>; height: number; width: number }

export const FOLDER_LABEL_TOO_SMALL_PARENT = 0.2
export const FOLDER_LABEL_TOO_SMALL_ROUTE = 0.02
export const INITIAL_PADDING = 50
export const DEFAULT_PADDING_FLOOR_LABEL_FROM_LEVEL_0 = 100
export const DEFAULT_PADDING_FLOOR_LABEL_FROM_LEVEL_1 = 85
export const PADDING_APPROX_FOR_DEPTH_ZERO = 0.035
export const PADDING_APPROX_FOR_DEPTH_ONE = 0.028

const MIN_BUILDING_AREA = 100
const PADDING_SCALING_FACTOR = 0.4

export function createTreemapNodes(map: CodeMapNode, state: State, metricData: NodeMetricData[], isDeltaState: boolean): Node[] {
	const mapSizeResolutionScaling = getMapResolutionScaleFactor(state.files)
	const maxHeight = metricData.find(x => x.name === state.dynamicSettings.heightMetric).maxValue * mapSizeResolutionScaling
	const maxWidth = metricData.find(x => x.name === state.dynamicSettings.areaMetric).maxValue * mapSizeResolutionScaling
	const heightScale = (treeMapSize * 2) / maxHeight
	if (hasFixedFolders(map)) {
		const scaledPadding = (INITIAL_PADDING * state.dynamicSettings.margin) / 100
		const hierarchyNode = hierarchy(map)
		// Base root folder has width: 100px and length: 100px
		const nodes: Node[] = [TreeMapHelper.buildRootFolderForFixedFolders(hierarchyNode.data, heightScale, state, isDeltaState)] // nosonar
		// Multiply mapSize of (default) 250px by 2 = 500px and add the total margin
		const totalMapSize = treeMapSize * 2 + getEstimatedNodesPerSide(hierarchyNode) * (scaledPadding / PADDING_SCALING_FACTOR)
		// than divide through the root folder width and length to get a scale factor for calculation for all following nodes.
		const scaleLength = totalMapSize / nodes[0].width
		const scaleWidth = totalMapSize / nodes[0].length
		// Scale the 100x100 root folder to be bigger and to match fixed/estimated totalMapSize
		scaleRoot(nodes[0], scaleLength, scaleWidth)
		return [
			...nodes,
			...buildSquarifiedTreeMapsForFixedFolders(
				hierarchyNode,
				state,
				scaleLength,
				scaleWidth,
				0,
				0,
				heightScale,
				maxHeight,
				maxWidth,
				isDeltaState,
				mapSizeResolutionScaling
			)
		]
	}

	const squarifiedTreeMap = getSquarifiedTreeMap(map, state)

	const nodes: Node[] = []
	for (const squarifiedNode of squarifiedTreeMap.treeMap) {
		nodes.push(TreeMapHelper.buildNodeFrom(squarifiedNode, heightScale, maxHeight, state, isDeltaState))
	}
	return nodes
}
function buildSquarifiedTreeMapsForFixedFolders(
	hierarchyNode: HierarchyNode<CodeMapNode>,
	state: State,
	scaleLength: number,
	scaleWidth: number,
	offsetX0: number,
	offsetY0: number,
	heightScale: number,
	maxHeight: number,
	maxWidth: number,
	isDeltaState: boolean,
	mapSizeResolutionScaling: number
) {
	const nodes = []

	for (const fixedFolder of hierarchyNode.children) {
		const fixedPosition = fixedFolder.data.fixedPosition
		const squarified = getSquarifiedTreeMap(fixedFolder.data, state)

		for (const squarifiedNode of squarified.treeMap.descendants()) {
			// squarified.width/height is a sum of the fixedPosition.width/height and applied margins
			// The following scaling factors are used later to calculate out the original percentage fixed width/height.
			// Example: Width and Height fixed at 20. Added margins 42px (absolute value) for each = 62px
			//  20/62 = 0.322...       62 x 0.322... = 20 :)
			const scaleX = fixedPosition.width / squarified.width
			const scaleY = fixedPosition.height / squarified.height
			// Scale the fixed (percentage) positions of a node by the right scale factor, so that it will be placed properly.
			// The squarifiedNode coordinates x0, x1, y0, y1 are already assigned with positions from the treemap algorithm.
			squarifiedNode.x0 = (squarifiedNode.x0 * scaleX + fixedPosition.left) * scaleWidth
			squarifiedNode.x1 = (squarifiedNode.x1 * scaleX + fixedPosition.left) * scaleWidth
			squarifiedNode.y0 = (squarifiedNode.y0 * scaleY + fixedPosition.top) * scaleLength
			squarifiedNode.y1 = (squarifiedNode.y1 * scaleY + fixedPosition.top) * scaleLength
			// Add x and y (absolute px) offsets from parent fixed folder, if any.
			squarifiedNode.x0 += offsetX0
			squarifiedNode.x1 += offsetX0
			squarifiedNode.y0 += offsetY0
			squarifiedNode.y1 += offsetY0
			const node = TreeMapHelper.buildNodeFrom(squarifiedNode, heightScale, maxHeight, state, isDeltaState, true)
			nodes.push(node)
			if (hasFixedFolders(fixedFolder.data)) {
				// Imagine the parent Folder has absolute px-width of 341px
				// We need to calculate a scale factor for the fixed child to transform a relative percentage fixed width/height to px.
				// Example: 20% of 341px:
				//  fixedWidth=20, parent Fixed Folder width: 341px => 20 * (scaleLength=341/100)
				// The original scaleWidth/scaleLength must be retained for the case that you have more than one fixedFolder on the same level.
				const childRelativeLengthScale = node.length / 100
				const childRelativeWidthScale = node.width / 100
				Array.prototype.push.apply(
					nodes,
					buildSquarifiedTreeMapsForFixedFolders(
						fixedFolder,
						state,
						childRelativeLengthScale,
						childRelativeWidthScale,
						squarifiedNode.x0,
						squarifiedNode.y0,
						heightScale,
						maxHeight,
						maxWidth,
						isDeltaState,
						mapSizeResolutionScaling
					)
				)
				// the break is actually needed!
				// the inner for-loop loops over the (parent) fixedFolder and its descendants.
				// if a direct child is a fixedFolder as well it will be handled by the recursive function call.
				// In this case, we must break the inner loop after handling the fixedFolder child recursively
				// and therefore prevent that the fixed child will be added as a node twice.
				break
			}
		}
	}
	return nodes
}
function hasFixedFolders(map: CodeMapNode) {
	// What if the second child of root is the first fixed folder?
	// Assumption: all children of the root folder require the fixedPosition attribute.
	return Boolean(map.children[0]?.fixedPosition)
}
function scaleRoot(root: Node, scaleLength: number, scaleWidth: number) {
	root.x0 *= scaleWidth
	root.y0 *= scaleLength
	root.width *= scaleWidth
	root.length *= scaleLength
}

export function calculateBuildingAreaWithProportionalPadding(area: number, smallestDelta: number, minimumBuildingArea: number) {
	if (area === 0) {
		return 0
	}

	return (area / smallestDelta) * minimumBuildingArea
}

export function getBuildingAreasWithProportionalPadding(childrenAreaValues: number[], smallestDelta: number, minimumBuildingSize: number) {
	return childrenAreaValues.map(element => {
		return calculateBuildingAreaWithProportionalPadding(element, smallestDelta, minimumBuildingSize)
	})
}

function getSquarifiedTreeMap(map: CodeMapNode, state: State): SquarifiedTreeMap {
	map.children = map.children.filter(child => child.attributes[state.dynamicSettings.areaMetric] !== 0)

	const hierarchyNode = hierarchy(map)
	let padding = INITIAL_PADDING * (state.dynamicSettings.margin / 100)

	// from hierarchy Node to areas
	const childrenAreaValues = getChildrenAreaValues(hierarchyNode, state)

	const smallestDelta = getSmallestValueOrSmallestDifference(childrenAreaValues)

	padding = calculatePaddingBasedOnBuildingArea(childrenAreaValues, smallestDelta, MIN_BUILDING_AREA, padding)

	const metricBuildingAreasIncludingPadding = getBuildingAreasWithProportionalPadding(
		childrenAreaValues,
		smallestDelta,
		MIN_BUILDING_AREA
	)

	const { rootWidth, rootHeight, metricSum } = calculateTotalNodeArea(metricBuildingAreasIncludingPadding, hierarchyNode, padding, state)

	const width = rootWidth
	const height = rootHeight
	let rootNode

	const treeMap = treemap<CodeMapNode>()
		.size([height, width]) // padding does not need to be divided by 2, already calculated by d3-hierarchy
		.paddingOuter(padding) // padding of the element; applied to leaves
		.paddingInner(padding) // margin of the element; applied to folders
		.paddingRight(node => {
			if (!rootNode && node.parent === null) {
				rootNode = node
			}
			// it seems that depth property is missing in that case
			// so the default padding will be added, which is fine though.
			if (rootNode) {
				const rootSize = rootNode.x1 - rootNode.x0

				// Start the labels at level 1 not 0 because the root folder should not be labeled
				if (node.depth === 0) {
					// Add a big padding for the first folder level (the font is bigger than in deeper levels)
					return Math.max(rootSize * PADDING_APPROX_FOR_DEPTH_ZERO, DEFAULT_PADDING_FLOOR_LABEL_FROM_LEVEL_0)
				}
				if (
					node.depth > 0 &&
					node.depth < 3 &&
					node.value / node.parent.value > FOLDER_LABEL_TOO_SMALL_PARENT &&
					node.value / rootNode.value > FOLDER_LABEL_TOO_SMALL_ROUTE
				) {
					return Math.max(rootSize * PADDING_APPROX_FOR_DEPTH_ONE, DEFAULT_PADDING_FLOOR_LABEL_FROM_LEVEL_1)
				}
			}

			// add treemap algorithm default padding otherwise
			return padding
		})

	const logSum = treeMap(metricSum)

	return { treeMap: logSum, height, width }
}

function getEstimatedNodesPerSide(hierarchyNode: HierarchyNode<CodeMapNode>) {
	let totalNodes = 0
	let blacklistedNodes = 0
	hierarchyNode.each(({ data }) => {
		if (data.isExcluded || data.isFlattened) {
			blacklistedNodes++
		}
		totalNodes++
	})
	// What does this line do?
	// Imagine a 3x3 grid of 9 nodes
	// 3 nodes are placed on the x-axis and 3 on the y-axis = 6
	// The calculated value is probably used to calculate the total margin which extends length and width of the map.
	return 2 * Math.sqrt(totalNodes - blacklistedNodes)
}
function isOnlyVisibleInComparisonMap(node: CodeMapNode, dynamicSettings: DynamicSettings) {
	return node.attributes[dynamicSettings.areaMetric] === 0 && node.deltas[dynamicSettings.heightMetric] < 0
}

// Only exported for testing.
export function getAreaValue(node: CodeMapNode, { dynamicSettings }: State) {
	if (node.isExcluded) {
		return 0
	}
	if (node.deltas && isOnlyVisibleInComparisonMap(node, dynamicSettings)) {
		return Math.abs(node.deltas[dynamicSettings.areaMetric])
	}

	if (isLeaf(node) && node.attributes?.[dynamicSettings.areaMetric]) {
		return node.attributes[dynamicSettings.areaMetric]
	}
	return 0
}
