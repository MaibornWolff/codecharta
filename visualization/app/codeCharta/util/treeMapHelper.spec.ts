import {TreeMapHelper} from "./treeMapHelper";
import {SquarifiedValuedCodeMapNode} from "./treeMapGenerator";
import {CodeMapNode, Settings} from "../codeCharta.model";
import {SETTINGS} from './dataMocks'

describe("treeMapHelper", () => {

    describe("build node", () => {

        let codeMapNode: CodeMapNode;
        let squaredNode: SquarifiedValuedCodeMapNode;
        let settings

        let heightScale = 1;
        let heightValue = 100;
        let depth = 0;
        let parent = null;
        let minHeight = 1;
        let folderHeight = 2;
        let maxHeight = 2000;

        beforeEach(() => {

            codeMapNode = {
                name: "Anode",
                path: "/root/Anode",
                type: "File",
                attributes: {}
            } as CodeMapNode;

            squaredNode = {
                data: codeMapNode,
                value: 42,
                x0: 0,
                y0: 0,
                x1: 400,
                y1: 400
            } as SquarifiedValuedCodeMapNode;

            settings = SETTINGS
            settings.treeMapSettings.mapSize = 1
            settings.dynamicSettings.margin = 15
            settings.appSettings.invertHeight = false

        });

        function buildNode() {
            return TreeMapHelper.buildNodeFrom(
                squaredNode,
                heightScale,
                heightValue,
                maxHeight,
                depth,
                parent,
                settings,
                minHeight,
                folderHeight
            );
        }

        it("minimal", () => {
            expect(buildNode()).toMatchSnapshot();
        });

        it("invertHeight", () => {
            settings.appSettings.invertHeight = true;
            expect(buildNode()).toMatchSnapshot();
        });

        it("deltas", () => {
            squaredNode.data.deltas = {};
            settings.dynamicSettings.heightMetric = "theHeight"
            squaredNode.data.deltas[settings.dynamicSettings.heightMetric] = 33;
            expect(buildNode()).toMatchSnapshot();
            squaredNode.data.deltas = undefined;
        });

        it("given negative deltas the resulting heightDelta also should be negative", () => {
            squaredNode.data.deltas = {};
            squaredNode.data.deltas[settings.dynamicSettings.heightMetric] = -33;
            expect(buildNode().heightDelta).toBe(-33);
            squaredNode.data.deltas = undefined;
        });

        it("is leaf", () => {
            let tmp = TreeMapHelper.isNodeLeaf;
            TreeMapHelper.isNodeLeaf = jest.fn(() => {return true});
            expect(buildNode()).toMatchSnapshot();
            TreeMapHelper.isNodeLeaf = tmp;
        });

        it("should set lowest possible height caused by other visible edge pairs", () => {
            settings.fileSettings.edges = [{
                fromNodeName: "/root/AnotherNode1",
                toNodeName: "/root/AnotherNode2",
                attributes: {
                    pairingRate: 33,
                    avgCommits: 12
                },
                visible: true,
            }];
            expect(buildNode()).toMatchSnapshot();
        });

        it("should set markingColor according to markedPackages", () => {
            const color = "#FF0000";
            settings.fileSettings.markedPackages = [{
                path: "/root/Anode",
                color: color,
                attributes: {},
            }];
            expect(buildNode().markingColor).toEqual(color);
        });

        it("should set no markingColor according to markedPackages", () => {
            const color = "#FF0000";
            settings.fileSettings.markedPackages = [{
                path: "/root/AnotherNode",
                color: color,
                attributes: {},
            }];
            expect(buildNode().markingColor).toEqual(null);
        });
    });

    describe("detect leaves", () => {

        it("root with no children is leaf", () => {
            const root = {};
            expect(TreeMapHelper.isNodeLeaf(root)).toBeTruthy();
        });

        it("root plus child should not be a leaf", () => {
            const root = {children: [{}]};
            expect(TreeMapHelper.isNodeLeaf(root)).toBeFalsy();
            expect(TreeMapHelper.isNodeLeaf(root.children[0])).toBeTruthy();
        });

        it("root plus child in child", () => {
            const root = {children: [{children: [{}]}]};
            expect(TreeMapHelper.isNodeLeaf(root)).toBeFalsy();
            expect(TreeMapHelper.isNodeLeaf(root.children[0])).toBeFalsy();
            expect(TreeMapHelper.isNodeLeaf(root.children[0].children[0])).toBeTruthy();
        });

        it("root plus two children", () => {
            const root = {children: [{}, {}]};
            expect(TreeMapHelper.isNodeLeaf(root)).toBeFalsy();
            expect(TreeMapHelper.isNodeLeaf(root.children[0])).toBeTruthy();
            expect(TreeMapHelper.isNodeLeaf(root.children[1])).toBeTruthy();
        });

    });

    describe("count nodes", () => {

        it("root only should be 1", () => {
            const root = {};
            expect(TreeMapHelper.countNodes(root)).toBe(1);
        });

        it("root plus child should be 2", () => {
            const root = {children: [{}]};
            expect(TreeMapHelper.countNodes(root)).toBe(2);
        });

        it("root plus child in child should be 3", () => {
            const root = {children: [{children: [{}]}]};
            expect(TreeMapHelper.countNodes(root)).toBe(3);
        });

        it("root plus two children should be 3", () => {
            const root = {children: [{}, {}]};
            expect(TreeMapHelper.countNodes(root)).toBe(3);
        });

    });

    describe("isNodeToBeFlat", () => {

        let codeMapNode: CodeMapNode;
        let squaredNode: SquarifiedValuedCodeMapNode;
        let treeMapSettings: Settings

        beforeEach(() => {

            codeMapNode = {
                name: "Anode",
                path: "/root/Anode",
                type: "File",
                attributes: {}
            } as CodeMapNode;

            squaredNode = {
                data: codeMapNode,
                value: 42,
                x0: 0,
                y0: 0,
                x1: 400,
                y1: 400
            } as SquarifiedValuedCodeMapNode;

            treeMapSettings = SETTINGS
            treeMapSettings.treeMapSettings.mapSize = 1
            treeMapSettings.dynamicSettings.margin = 15

        });

        it("should not be a flat node when no visibleEdges", () => {
            treeMapSettings.fileSettings.edges = [];
            expect(TreeMapHelper["isNodeToBeFlat"](squaredNode, treeMapSettings)).toBeFalsy();
        });

        it("should be a flat node when other edges are visible", () => {
            treeMapSettings.fileSettings.edges = [{
                fromNodeName: "/root/anotherNode",
                toNodeName: "/root/anotherNode2",
                attributes: {},
                visible: true
            }];
            expect(TreeMapHelper["isNodeToBeFlat"](squaredNode, treeMapSettings)).toBeTruthy();
        });

        it("should not be a flat node when it contains edges", () => {
            treeMapSettings.fileSettings.edges = [{
                fromNodeName: "/root/Anode",
                toNodeName: "/root/anotherNode",
                attributes: {}
            }];
            expect(TreeMapHelper["isNodeToBeFlat"](squaredNode, treeMapSettings)).toBeFalsy();
        });

        it("should not be a flat node, because its searched for", () => {
            treeMapSettings.dynamicSettings.searchedNodePaths = ["/root/Anode"];
            treeMapSettings.dynamicSettings.searchPattern = "Anode";
            expect(TreeMapHelper["isNodeToBeFlat"](squaredNode, treeMapSettings)).toBeFalsy();
        });

        it("should be a flat node, because other nodes are searched for", () => {
            treeMapSettings.dynamicSettings.searchedNodePaths = ["/root/anotherNode", "/root/anotherNode2"];
            treeMapSettings.dynamicSettings.searchPattern = "Anode";
            expect(TreeMapHelper["isNodeToBeFlat"](squaredNode, treeMapSettings)).toBeTruthy();
        });

        it("should not be a flat node when searchPattern is empty", () => {
            treeMapSettings.dynamicSettings.searchedNodePaths = ["/root/anotherNode", "/root/anotherNode2"];
            treeMapSettings.dynamicSettings.searchPattern = "";
            expect(TreeMapHelper["isNodeToBeFlat"](squaredNode, treeMapSettings)).toBeFalsy();
        });

    });

});