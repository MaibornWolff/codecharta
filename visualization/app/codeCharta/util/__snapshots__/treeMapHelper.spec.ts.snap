// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`treeMapHelper build node deltas 1`] = `
Object {
  "attributes": Object {},
  "children": Array [],
  "deltas": Object {
    "theHeight": 33,
  },
  "depth": 0,
  "flat": false,
  "height": 100,
  "heightDelta": 33,
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": null,
  "name": "Anode",
  "origin": undefined,
  "parent": null,
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`treeMapHelper build node invertHeight 1`] = `
Object {
  "attributes": Object {},
  "children": Array [],
  "deltas": undefined,
  "depth": 0,
  "flat": false,
  "height": 1900,
  "heightDelta": 0,
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": null,
  "name": "Anode",
  "origin": undefined,
  "parent": null,
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`treeMapHelper build node is leaf 1`] = `
Object {
  "attributes": Object {},
  "children": Array [],
  "deltas": undefined,
  "depth": 0,
  "flat": false,
  "height": 100,
  "heightDelta": 0,
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": null,
  "name": "Anode",
  "origin": undefined,
  "parent": null,
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`treeMapHelper build node minimal 1`] = `
Object {
  "attributes": Object {},
  "children": Array [],
  "deltas": undefined,
  "depth": 0,
  "flat": false,
  "height": 100,
  "heightDelta": 0,
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": null,
  "name": "Anode",
  "origin": undefined,
  "parent": null,
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`treeMapHelper build node should set lowest possible height caused by other visible edge pairs 1`] = `
Object {
  "attributes": Object {},
  "children": Array [],
  "deltas": undefined,
  "depth": 0,
  "flat": true,
  "height": 1,
  "heightDelta": 0,
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": null,
  "name": "Anode",
  "origin": undefined,
  "parent": null,
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;
